README.md
## Table of Contents

1. [Usage](#usage)
   1. [GFF3, GVF, VCF to JSON](#gff3-gvf-vcf-to-json)
   2. [JSON to GFF3, GVF, VCF](#json-to-gff3-gvf-vcf)
   3. [Python API](#python-api)
   4. [RethinkDB](#rethinkdb)
   5. [MongoDB](#mongodb)
2. [Data Model](#data-model)
   1. [Overview](#overview)
      1. [Context Objects](#context-objects)
      2. [Meta Objects](#meta-objects)
      3. [Feature Objects](#feature-objects)
      4. [Summary Objects](#summary-objects)
3. [Python API Reference Cards](#python-api-reference-cards)
4. [JSON Reference Cards](#json-reference-cards)

## Usage

BioInterchange is a command line tool. On OS X, it can be run via the [Terminal](https://en.wikipedia.org/wiki/Terminal_(OS_X)) application. On Linux, well, you use Linux, so you know what a command line is.

Running BioInterchange will cause the software to perform a quick system check before anything else happens. If there are incompatibility problems, then these will be reported and the software exits.

BioInterchange is open source now and free to use. If your version of BioInterchange complains about a missing licensing file, then update your BioInterchange installation to the latest verion.

<div class="row">
    <div class="col-xs-12 col-sm-10 col-md-10 col-lg-10 col-xs-offset-0 col-sm-offset-1 col-md-offset-1 col-lg-offset-1">
        <div class="panel panel-info" role="alert">
            <div class="panel-heading">
                Note
            </div>
            <div class="panel-body">
                Trial licenses and licenses for reviewers in the peer-review process are free.
            </div>
        </div>
    </div>
</div>

##### Abbreviations

<dl class="dl-horizontal">
    <dt>GFF3</dt><dd><a href="http://www.sequenceontology.org/gff3.shtml">Generic Feature Format Version 3</a></dd>
    <dt>GVF</dt><dd><a href="http://www.sequenceontology.org/resources/gvf.html">Genome Variation Format</a></dd>
    <dt>JSON</dt><dd><a href="http://json.org/">JavaScript Object Notation</a></dd>
    <dt>JSON-LD</dt><dd><a href="http://json-ld.org/">JSON Linked Data</a></dd>
    <dt>LDJ/LDJSON</dt><dd><a href="https://en.wikipedia.org/wiki/Line_Delimited_JSON">Line Delimited JSON</a></dd>
    <dt>VCF</dt><dd><a href="https://samtools.github.io/hts-specs/VCFv4.2.pdf">Variant Call Format</a></dd>
</dl>

### GFF3, GVF, VCF to JSON

Converting genomic file formats to a series of JSON objects:

    biointerchange -o example.ldj example.vcf

Here, the genomics file "example.vcf" is the input. The output of BioInterchange will be written to the file "example.ldj". The "-o" parameter can also be omitted, in which case the output is written directly on the console:

    biointerchange example.vcf

The optional "-u" parameter can be used to add a custom user annotation to the [context object](#context-objects). This can be useful when including BioInterchange in a genomics analysis pipeline:

    biointerchange -u "CNV Analysis; Smith Lab" -o cnv.ldj cnv.gvf

BioInterchange prints its semantic version number with the "-v" parameter:

    biointerchange -v

To show the EULA that came with the software, the "-e" parameter comes in handy:

    biointerchange -e

A brief help text is shown with the "-h" parameter:

    biointerchange -h

### JSON to GFF3, GVF, VCF

Converting JSON objects back to their original genomic file format:

    biointerchange -o example.vcf example.ldj

<div class="row">
    <div class="col-xs-12 col-sm-10 col-md-10 col-lg-10 col-xs-offset-0 col-sm-offset-1 col-md-offset-1 col-lg-offset-1">
        <div class="panel panel-info" role="alert">
            <div class="panel-heading">
                Note
            </div>
            <div class="panel-body">
                Modified JSON objects (for example, when using the Python API) will only translate back correctly when non-standard key/value pairs are put under the &ldquo;user-defined&rdquo; key.
            </div>
        </div>
    </div>
</div>

### Python API

Genomics data can be accessed and modified via the Python API. Each JSON object will be passed on to a Python function and changes made by the Python code will be preserved:

    biointerchange -p simplepy.simple test-data/playground.gff3

<div class="row">
    <div class="col-xs-12 col-sm-10 col-md-10 col-lg-10 col-xs-offset-0 col-sm-offset-1 col-md-offset-1 col-lg-offset-1">
        <div class="panel panel-info" role="alert">
            <div class="panel-heading">
                Compatibility
            </div>
            <div class="panel-body">
                Requires Python 3.4 or newer.
            </div>
        </div>
    </div>
</div>

<dl>
    <dt>Example: Directory structure.</dt>
    <dd markdown="1">
* ./simplepy/\_\_init\_\_.py
* ./simplepy/simple.py
</dd>
    <dt>Example: Environment variables.</dt>
    <dd markdown="1">
* PYTHONHOME needs to be set to where Python is installed on your system
* PYTHONPATH needs to include "." for this example, so that "simple.py" can be found
</dd>
    <dt>Example: &ldquo;simple.py&rdquo; source code.</dt>
    <dd markdown="1">
~~~ python
# This variable will be used to accumulate the
# length (in basepairs) of all features that are
# not filtered out in `process_feature` below.
#
# Use this or a similar approach to share data
# across features.
__accumulated_length__ = 0

def setup(context, meta):
    # The context is always being output, but meta
    # could be omitted. How? It is quite simple:
    # returning (a possibly modified) meta Dict
    # will be output, but returning None will
    # output no meta information.
    return meta

def cleanup(summary):
    # Add the accumulated length of all features
    # to the summary:
    summary['my-user-data'] = { 'accumulated-length' : __accumulated_length__ }

    # The Dict `summary` needs to be returned in order
    # to appear in the output. If no summary should appear
    # in the output, then None needs to be returned.
    return summary

def process_feature(feature):
    global __accumulated_length__

    locus = feature['locus']

    length = abs(locus['end'] - locus['start']) + 1

    # If we want to filter out some features (remove them
    # from the output completely), then we can do so my
    # returning None.
    if length < 10000:
        return None

    __accumulated_length__ += length

    feature['my-calculated-length'] = length

    if 'comment' in feature:
        del feature['comment']

    return feature
~~~
</dd>
</dl>

### RethinkDB

Take the cat lovers example from the main page:

    wget ftp://ftp.ensembl.org/pub/release-81/variation/vcf/felis_catus/Felis_catus_incl_consequences.vcf.gz
    gunzip Felis_catus_incl_consequences.vcf.gz
    biointerchange -o Felis_catus_incl_consequences.ldj Felis_catus_incl_consequences.vcf

Import line-delimited JSON-LD documents into RethinkDB:

    rethinkdb import -f Felis_catus_incl_consequences.ldj --table genomics.felis_catus

Check that the data is actually in the database using RethinkDB's "Data Explorer" (for local installation at `http://localhost:8080/#dataexplorer`):

    r.db('genomics').table('felis_catus')

Voila!

### MongoDB

<div class="row">
    <div class="col-xs-12 col-sm-10 col-md-10 col-lg-10 col-xs-offset-0 col-sm-offset-1 col-md-offset-1 col-lg-offset-1">
        <div class="panel panel-info" role="alert">
            <div class="panel-heading">
                Note
            </div>
            <div class="panel-body">
                MongoDB can be a bit much to begin with. Have a look at RethinkDB for starters! RethinkDB has a great intuitive web-interface, easy-peasy indexing and sharding support, and a cool query language (ReQL).
            </div>
        </div>
    </div>
</div>

Take the cat lovers example from the main page:

    wget ftp://ftp.ensembl.org/pub/release-81/variation/vcf/felis_catus/Felis_catus_incl_consequences.vcf.gz
    gunzip Felis_catus_incl_consequences.vcf.gz
    biointerchange -o Felis_catus_incl_consequences.ldj Felis_catus_incl_consequences.vcf

Import line-delimited JSON-LD documents into MongoDB:

    mongoimport --db genomics --collection felis_catus --type json --file Felis_catus_incl_consequences.ldj

Check that the data is actually in the database:

    mongo genomics
    > db.felis_catus.find()

Voila!

## Data Model

### Overview

Four kind of JSON objects are created with each execution of BioInterchange -- in this particular order:

1. a "context object"
2. zero, one, or more "meta objects" (pragma and information lines)
3. a series of "feature objects" (the actual genomics feature-data)
4. a "summary object"

Relationships between objects:

* context and summary objectss stand on their own; they are describing how the data was created, how much data was processed, and how long it took
* meta objects themselves are also independent, but they can be referenced (linked to) by feature objects
* feature objects can reference meta objects as well as other feature objects

<div class="row">
    <div class="col-xs-12 col-sm-10 col-md-10 col-lg-10 col-xs-offset-0 col-sm-offset-1 col-md-offset-1 col-lg-offset-1">
        <div class="panel panel-info" role="alert">
            <div class="panel-heading">
                Note
            </div>
            <div class="panel-body">
                In a nutshell: meta objects and feature objects contain the genomics data from GFF3, GVF, and VCF files; these objects are referencing each other (they are linked). Context and summary objects stand on their own.
            </div>
        </div>
    </div>
</div>

**All** objects contain a "@context" key and a "@type" key, which are called context key and type key in the following.

The context key turns the JSON objects into JSON-LD objects. Never heard of JSON-LD? Just skip to the next part and ignore the "@context" key. This is what makes JSON-LD so great: you can handle JSON-LD objects just like JSON objects.

If you do want to make use of the context key, then feed the JSON-LD objects to a Linked Data tool and it will annotate key/value pairs with type information. JSON-LD objects can also be turned into Triple Store compatible data formats, such as RDF N-Triples, RDF N-Quads, and RDF Turtle. Want to see a real-life example of this magic? Head over to the [JSON-LD Playground](http://json-ld.org/playground/) and copy/paste any JSON-LD object into the "JSON-LD Input" text field: the "N-Quads" tab will instantly show a triples representation of the JSON-LD object that can be loaded into a triple store.

<div class="row">
    <div class="col-xs-12 col-sm-10 col-md-10 col-lg-10 col-xs-offset-0 col-sm-offset-1 col-md-offset-1 col-lg-offset-1">
        <div class="panel panel-info" role="alert">
            <div class="panel-heading">
                Note
            </div>
            <div class="panel-body">
                All JSON-LD annotations are making use of GFVO<sup>2</sup>. Is it legit? Well, it is based on all the lessons learned when designing the <a href="https://peerj.com/articles/933/">Genomic Feature and Variation Ontology</a> (<a href="https://peerj.com/articles/933/">GFVO</a>). So, there is a team behind GFVO<sup>2</sup> too? Yes, stay tuned&hellip;
            </div>
        </div>
    </div>
</div>

#### Context Objects

Context objects tell you something about the environment in which the following JSON objects ("meta objects", "feature objects" and "summary objects") were created.

In essence, a context objects contains information about:

* which version of BioInterchange was used
* the input filename
* the filetype of the input (GFF3, GVF, VCF)
* what the output filename was (unless output went to the console)
* whether the Python API was utilized (if so, name of the Python module)
* additional user-defined parameters

<dl>
    <dt>Example</dt>
    <dd markdown="1">
    {
        "@context" : "https://www.codamono.com/jsonld/vcf-c1.json",
        "biointerchange-version" : "2.0.0+36",
        "input-filetype" : "VCF",
        "output-file" : null,
        "input-file" : "test-data/playground-vcf.vcf",
        "python-callback" : null,
        "user-defined" : null
    }
</dd>
</dl>

Detailed information about each key/value-pair can be found in the JSON Reference Card section.

#### Meta Objects

For each genomic file, one meta object is being created. The meta object contain data of information/pragma lines.

Meta objects are helpful for determining data provenance, establishing links to ontologies that were used, and providing extra annotations that were not attached to features to reduce data redundancy.

For example, the following example provides a textual descriptions for analytic filters that have been applied in a VCF file. Instead of adding this description to every genomic feature for which the filter applies, the textual description is only given once in this meta object, where it can be looked up via the filter keys ("MinAB", "MinDP", "MinMQ", and "Qual").

<dl>
    <dt>Example:</dt>
    <dd markdown="1">
    {
        "@context" : "https://www.codamono.com/jsonld/vcf-x1.json",
        "vcf-version" : "4.2",
        "reference" : "ftp://ftp-mouse.sanger.ac.uk/ref/GRCm38_68.fa",
        "contig" : {
            "1" : {
                "length" : 195471971
            },
            "3" : {
                "length" : 918278173
            }
        },
        "filter" : {
            "MinAB" : {
                "description" : "Minimum number of alternate bases (INFO/DP4) [5]"
            },
            "MinDP" : {
                "description" : "Minimum read depth (INFO/DP or INFO/DP4) [5]"
            },
            "MinMQ" : {
                "description" : "Minimum RMS mapping quality for SNPs (INFO/MQ) [20]"
            },
            "Qual" : {
                "description" : "Minimum value of the QUAL field [10]"
            }
        },
        "user-defined" : {
            "samtoolsVersion" : [
                "0.1.18-r572"
            ]
        }
    }
</dd>
</dl>

#### Feature Objects

Sequences, sequence variations, sequence annotations, genotyping samples, etc., are represented by feature objects.

##### Basic Feature Information

Most basic information about features includes an identifier, a genomic locus, provenane ("SGRP" stands for the Saccharomyces Genome Resequencing Project), feature type information ("SNV" -- single nucleotide variant; a Sequence Ontology term), and references to external databases (SGRP, again, and European Molecular Biology Laboratory accession).

<dl>
    <dt>Example</dt>
    <dd markdown="1">
    {
        "@context" : "https://www.codamono.com/jsonld/gvf-f1.json",
        "id" : "76",
        "locus" : {
            "landmark" : "Chr1",
            "start" : 675,
            "end" : 675,
            "strand" : "+"
        },
        "source" : "SGRP",
        "type" : "SNV",
        "dbxref" : [
            "SGRP:s01-675",
            "EMBL:AA816246"
        ]
    }
</dd>
</dl>

##### Variant and Reference Sequences

Information about reference sequences is stored under the "reference" key. Variations are stored under the "variants" key and allele specific informations is labeled by "B", "C", etc. ("A" is denoting the reference, which is not explicitly labeled).

<dl>
    <dt>Example: GVF (basic information omitted)</dt>
    <dd markdown="1">
    {
        "@context" : "https://www.codamono.com/jsonld/gvf-f1.json",
        "reference" : {
            "sequence" : "A",
            "codon" : "GAG"
        },
        "variants" : {
            "B" : {
                "sequence" : "G",
                "codon" : "GAG"
            },
            "C" : {
                "codon" : "GGG",
                "sequence" : "T"
            }
        }
    }
</dd>

templates/changelog
biointerchange (VERSION)

 * TODO

-- Joachim Baran <kim@codamono.com> DATE

biointerchange (2.0.4+105)

 * Update: improved file access
 * Bugfix: memory leak that affected features with little/no data

-- Joachim Baran <kim@codamono.com> Thu Sep 10 16:39:05 UTC 2015

biointerchange (2.0.3+100)

 * Update: VCF 4.3 support
 * Update: better support for samples in VCF
 * Update: "-q" parameter for faster GFF3/GVF loading
 * Bugfix: corrected wrong/missing VCF key/value pairs
 * Bugfix: better handling of locus ranges in GVF
 * Bugfix: sequence region serialization for GFF3/GVF
 * Bugfix: genotype field bugfix when converting LDJ to VCF
 * Bugfix: better special character handling 

-- Joachim Baran <kim@codamono.com> Mon Aug 24 02:23:18 UTC 2015

biointerchange (2.0.2+84)

 * Bugfix: Memory leak that affected reading GVF/VCF and LDJ.
 * Update: License clarifications; reverse engineering clause.

-- Joachim Baran <kim@codamono.com> Sun Aug 16 02:57:18 UTC 2015

biointerchange (2.0.1+75)

 * Bugfix: Python reference counting.

-- Joachim Baran <kim@codamono.com> Wed Jul 29 13:06:25 UTC 2015

biointerchange (2.0.0+71)

 * Initial Release.

-- Joachim Baran <kim@codamono.com> Wed Jul 29 13:06:25 UTC 2015



include/gen.h
/*
 * Copyright (c) 2015 CODAMONO, Ontario, Canada
 *
 * This Source Code Form is subject to the terms of the accompanying
 * LICENSE.txt file, or, available via the following URLs:
 *
 *   As text: http://www.codamono.com/license/biointerchange-l1.txt
 *
 *   As PDF:  http://www.codamono.com/license/biointerchange-l1.pdf
 *
 *   As DOCX: http://www.codamono.com/license/biointerchange-l1.docx
 */

#ifndef biointerchange_gen_h
#define biointerchange_gen_h

#include <stdbool.h>
#include <stdint.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <unistd.h>

#include <document.h>
#include <json.h>

#include "ext-python.h"
#include "fio.h"

#ifdef __cplusplus
extern "C" {
#endif
    
// Error codes (exit codes):
#define MAIN_SUCCESS  0
    
// Never change these definitions or it will be very hard to interpret
// error codes across software versions!
#define MAIN_ERR_PGSZ 1
#define MAIN_ERR_HME1 2
#define MAIN_ERR_HME2 3
#define MAIN_ERR_PARA 4
// License file size wrong:
#define MAIN_ERR_LISZ 5
// License file format wrong:
#define MAIN_ERR_LICF 6
// License check network error:
#define MAIN_ERR_LICN 7
// License was rejected by the license server:
#define MAIN_ERR_LICV 8
#define MAIN_ERR_FNME 9
#define MAIN_ERR_FEXT 10
#define MAIN_ERR_FACC 11
// PEM_read_bio_X509 failed:
#define MAIN_ERR_SBIO 12
// X509_STORE_add_cert failed:
#define MAIN_ERR_SADD 13
// Format errors:
#define MAIN_ERR_GFMT_ATTR 14
#define MAIN_ERR_JFMT_KY 15
#define MAIN_ERR_JFMT_KYMAP 16
#define MAIN_ERR_JFMT_VAL 17
#define MAIN_ERR_JFMT_DEP 18
// Python errors:
#define MAIN_ERR_PYMD 19
#define MAIN_ERR_PYMEM 20
// System errors:
#define MAIN_ERR_SYSMMAP 21
#define MAIN_ERR_SYSMALL 22
#define MAIN_ERR_FLE 23
// Format errors:
#define MAIN_ERR_FMT 24
// Internal errors:
#define MAIN_ERR_INT 25
    
    // Internal error keys:
    //   A10293
    //   A76412
    //   A98613
    
// Ensure that no statistics are sent during license check:
#define GEN_STATS_PRIVATE 1
    
// Number of memory pages that should reside in memory when reading a genomics file:
#define BI_GEN_PG_MUL 4096

extern const char* JSONLD_CTX;
extern const char* JSONLD_TPE;

extern const char* JSONLD_CLSS_CTX;
extern const char* JSONLD_CLSS_MTA;
extern const char* JSONLD_CLSS_FTR;
extern const char* JSONLD_CLSS_SUM;
    
extern const char* JSONLD_GFF3_CTX1;
extern const char* JSONLD_GTF_CTX1;
extern const char* JSONLD_GVF_CTX1;
extern const char* JSONLD_VCF_CTX1;

extern const char* JSONLD_GFF3_X1;
extern const char* JSONLD_GTF_X1;
extern const char* JSONLD_GVF_X1;
extern const char* JSONLD_VCF_X1;

extern const char* JSONLD_GFF3_1;
extern const char* JSONLD_GTF_1;
extern const char* JSONLD_GVF_1;
extern const char* JSONLD_VCF_1;
    
extern const char* JSONLD_STAT_1;
    
extern const char* GEN_AFFECTED;
extern const char* GEN_AFFECTED_TPE;
extern const char* GEN_ALIAS;
extern const char* GEN_ALIAS_GFF3;
extern const char* GEN_ALIGNMENT;
extern const char* GEN_ALIGNMENT_GFF3;
extern const char* GEN_ALLELE_CNT;
extern const char* GEN_ALLELE_CNT_VCF;
extern const char* GEN_ALLELE_CNTEXP;
extern const char* GEN_ALLELE_CNTEXP_VCF;
extern const char* GEN_ALLELE_FRQ;
extern const char* GEN_ALLELE_FRQ_VCF;
extern const char* GEN_ALLELE_DEPTH;
extern const char* GEN_ALLELE_DEPTH_VCF;
extern const char* GEN_ALLELE_DEPTHFWD;
extern const char* GEN_ALLELE_DEPTHFWD_VCF;
extern const char* GEN_ALLELE_DEPTHREV;
extern const char* GEN_ALLELE_DEPTHREV_VCF;
extern const char* GEN_ALLELE_TTL;
extern const char* GEN_ALLELE_TTL_VCF;
extern const char* GEN_ANCESTRAL_ALLELE;
extern const char* GEN_ANCESTRAL_ALLELE_VCF;
extern const char* GEN_ANNOTATIONS;
extern const char* GEN_ANNOTATIONS_VCF;
extern const char* GEN_ATTRIBUTE_MTHD;
extern const char* GEN_ATTRS;
extern const char* GEN_AVG_COVERAGE;
extern const char* GEN_AVG_COVERAGE_GVF;
extern const char* GEN_BUILD;
extern const char* GEN_BUILD_GFF3;
extern const char* GEN_BUILD_VAL;
extern const char* GEN_CIGAR;
extern const char* GEN_CIGAR_GFF3;
extern const char* GEN_CIGAR_VCF;
extern const char* GEN_CODON;
extern const char* GEN_COMMENT;
extern const char* GEN_COMMENT_GVF;
extern const char* GEN_DATA_SRC;
extern const char* GEN_DATA_TPE;
extern const char* GEN_DATA_TPE_GVF;
extern const char* GEN_DBXREF;
extern const char* GEN_DBXREF_GFF3;
extern const char* GEN_DEPTH;
extern const char* GEN_DEPTH_VCF;
extern const char* GEN_EFFECT;
extern const char* GEN_EFFECTS;
extern const char* GEN_END;
extern const char* GEN_END_RANGE;
extern const char* GEN_END_RANGE_GVF;
extern const char* GEN_FILE_DATE;
extern const char* GEN_FILE_DATE_VCF1;
extern const char* GEN_FILE_DATE_VCF2;
extern const char* GEN_FMT_ALT;
extern const char* GEN_FMT_ALT_VCF;
extern const char* GEN_FMT_INFO;
extern const char* GEN_FMT_INFO_VCF;
extern const char* GEN_FMT_FLT;
extern const char* GEN_FMT_FLT_VCF;
extern const char* GEN_FMT_GT;
extern const char* GEN_FMT_GT_VCF;
extern const char* GEN_FST_BKPNT;
extern const char* GEN_FST_BKPNT_VCF;
extern const char* GEN_FST_REF;
extern const char* GEN_FST_REF_GVF;
extern const char* GEN_FST_REF_VCF;
extern const char* GEN_GENOMIC_SRC;
extern const char* GEN_GENOTYPE;
extern const char* GEN_GENOTYPE_VCF;
extern const char* GEN_GENOTYPE_LIKE;
extern const char* GEN_GENOTYPE_LIKE_VCF;
extern const char* GEN_GENOTYPE_LIKEP;
extern const char* GEN_GENOTYPE_LIKEP_VCF;
extern const char* GEN_GENOTYPE_QUAL;
extern const char* GEN_GENOTYPE_QUAL_VCF;
extern const char* GEN_GENOTYPE_PROB;
extern const char* GEN_GENOTYPE_PROB_VCF;
extern const char* GEN_GFFVERSION;
extern const char* GEN_GFFVERSION_GFF3;
extern const char* GEN_GLOBAL;
extern const char* GEN_GVFVERSION;
extern const char* GEN_GVFVERSION_GVF;
extern const char* GEN_HAP_QUALITIES;
extern const char* GEN_HAP_QUALITIES_VCF;
extern const char* GEN_ID;
extern const char* GEN_ID_GFF3;
extern const char* GEN_INDIVIDUALS;
extern const char* GEN_INDIVIDUALS_GVF1;
extern const char* GEN_INDIVIDUALS_GVF2;
extern const char* GEN_LANDMARK;
extern const char* GEN_LANDMARKS;
extern const char* GEN_LANDMARKS_GVF;
extern const char* GEN_LOCUS;
extern const char* GEN_MEMBER_1000G;
extern const char* GEN_MEMBER_1000G_VCF;
extern const char* GEN_MEMBER_DBSNP;
extern const char* GEN_MEMBER_DBSNP_VCF;
extern const char* GEN_MEMBER_HM2;
extern const char* GEN_MEMBER_HM2_VCF;
extern const char* GEN_MEMBER_HM3;
extern const char* GEN_MEMBER_HM3_VCF;
extern const char* GEN_ONT_ACCESSION;
extern const char* GEN_ONT_TERM;
extern const char* GEN_PHASE_SET;
extern const char* GEN_PHASE_SET_VCF;
extern const char* GEN_PHASE_QUAL;
extern const char* GEN_PHASE_QUAL_VCF;
extern const char* GEN_PHASED_GENO;
extern const char* GEN_PHENO_DESCR;
extern const char* GEN_POP;
extern const char* GEN_QUALITY_MAP;
extern const char* GEN_QUALITY_MAP_VCF;
extern const char* GEN_QUALITY_MAP0;
extern const char* GEN_QUALITY_MAP0_VCF;
extern const char* GEN_QUALITY_RMS;
extern const char* GEN_QUALITY_RMS_VCF;
extern const char* GEN_READ_PAIR_SPAN;
extern const char* GEN_READ_PAIR_SPAN_GVF;
extern const char* GEN_REFERENCE;
extern const char* GEN_SAMPLES_DATA;
extern const char* GEN_SAMPLES_DATA_VCF;
extern const char* GEN_SCORE_MTHD;
extern const char* GEN_SEQUENCE;
extern const char* GEN_SEQUENCE_REGION;
extern const char* GEN_SEQUENCE_REGION_GFF3;
extern const char* GEN_SEQUENCE_REGION_VCF;
extern const char* GEN_SEQUENCES;
extern const char* GEN_SEX;
extern const char* GEN_START;
extern const char* GEN_START_RANGE;
extern const char* GEN_START_RANGE_GVF;
extern const char* GEN_STRAND;
extern const char* GEN_STRAND_BIAS;
extern const char* GEN_STRAND_BIAS_VCF;
extern const char* GEN_SOMATIC;
extern const char* GEN_SOMATIC_VCF;
extern const char* GEN_SOURCE;
extern const char* GEN_SOURCES;
extern const char* GEN_SOURCES_GVF;
extern const char* GEN_SOURCE_MTHD;
extern const char* GEN_TECHNOLOGY;
extern const char* GEN_TYPE;
extern const char* GEN_TYPE_GVF;
extern const char* GEN_TYPES;
extern const char* GEN_TYPES_GVF;
extern const char* GEN_VALIDATED;
extern const char* GEN_VALIDATED_VCF;
extern const char* GEN_VARIANTS;
extern const char* GEN_VCFVERSION;
extern const char* GEN_VCFVERSION_VCF;

extern const char* GEN_EMPTY;
extern const char* GEN_NULL;
extern const char* GEN_TRUE;
extern const char* GEN_UNKNOWN;
extern const char* GEN_FORWARD;
    
// Statistics wrap-up:
extern const char* GEN_STATS;
extern const char* GEN_STAT_CMMS;
extern const char* GEN_STAT_META;
extern const char* GEN_STAT_META_FLTR;
extern const char* GEN_STAT_FTRS;
extern const char* GEN_STAT_FTRS_FLTR;
extern const char* GEN_STAT_RNTM;
extern const char* GEN_STAT_INVC;
extern const char* GEN_STAT_FNSH;
extern const char* GEN_STAT_LSEC;
    
extern ldoc_vis_nde_ord_t* json_vis_nde;;
extern ldoc_vis_ent_t* json_vis_ent;
    
// Letters in the alphabet; restriction for labelling alleles (A, B, C, etc.):
#define GEN_MAX_ALT       26
// Two characters needed to encode for the value '26':
#define GEN_MAX_ALT_CHARS  2
    
// Characters to step over when encoding for allele labels:
// Note -- assumes VCF_MAX_ALT <= 26, so that only one character is
//         required to encode for the allele.
#define GEN_STEP 3
    
extern char GEN_ALLELE[GEN_MAX_ALT * 2];
extern char GEN_ALLELES[GEN_STEP * (((GEN_MAX_ALT * (GEN_MAX_ALT + 1) / 2) + GEN_MAX_ALT + 1))];
    
typedef enum
{
    /**
     * Unknown keyword ("no keyword").
     */
    BI_NKW = 0,
    /**
     * Ignore. This key/value pair has been processed already.
     */
    BI_IGN,
    /**
     * String value.
     */
    BI_VAL,
    /**
     * Number value.
     */
    BI_NUM,
    /**
     * Boolean truth value (keyword indicates truth).
     */
    BI_BL,
    /**
     * Comma separated values.
     */
    BI_CSEP,
    /**
     * Comma separated values of pairs (colon separated):
     */
    BI_CSEPCPAIR,
    /**
     * Comma separated information about variants.
     */
    BI_CSEPVAR,
    /**
     * Comma separated information about variants that is a number.
     */
    BI_CSEPVARN,
    /**
     * Comma separated information about variants & skip 8 characters of the key.
     */
    BI_CSEPVAR8,
    /**
     * GVF "Variant_effect" key.
     */
    BI_GVFEFFECT,
    /**
     * Information about the reference sequence.
     */
    BI_REFSEQ,
    /**
     * Information about the reference sequence that is a number.
     */
    BI_REFSEQN,
    /**
     * Information about the reference sequence & skip 10 characters of the key.
     */
    BI_REFSEQ10,
    /**
     * CIGAR string in GFF3 format (needs reformatting to match actual CIGAR specification).
     */
    BI_XCIG,
    /**
     * Target string in GFF3 format.
     */
    BI_TARGET,
    BI_STRUCT
} bi_attr;
    
typedef struct gen_attr_t
{
    bi_attr attr;
    const char* alt;
} gen_attr_t;

typedef enum
{
    GEN_FMT_GFF3,
    GEN_FMT_GTF,
    GEN_FMT_GVF,
    GEN_FMT_VCF,
    GEN_FMT_LDJ
} gen_filetype_t;

typedef enum
{
    GEN_FMT_CTX,
    GEN_FMT_INF,
    GEN_FMT_FTR,
    GEN_FMT_STT
} gen_doctype_t;
    
typedef enum
{
    GEN_CTPE_SETUP,
    GEN_CTPE_CLEANUP,
    GEN_CTPE_PROCESS
} gen_ctpe_t;
    
typedef enum
{
    GEN_ALT_GL,
    GEN_ALT_GP,
    GEN_ALT_PL
} gen_alt_t;

typedef struct gen_fstat
{
    /**
     * Number of meta-information lines.
     */
    uint32_t meta;
    /**
     * Number of feature lines.
     */
    uint32_t ftrs;
    /**
     * Number of comment lines.
     */
    uint32_t comms;
    /**
     * Indicates whether the meta information was filtered.
     */
    bool cbk_meta_fltr;
    /**
     * Number of filtered feature lines.
     */
    uint32_t cbk_ftrs_fltr;
    /**
     * Offset to which the statistics are valid.
     */
    off_t off;
} gen_fstat;

typedef struct gen_fa_ntry_t
{
    off_t off;
    off_t seq;
    size_t llen;
    size_t lbrk;
} gen_fa_ntry_t;

/**
 * @brief Parser state.
 */
typedef struct gen_prsr_t
{
    bool fa_sct;
    bool vcf_ftr_sct;
    size_t vcf_col;
    char* vcf_hdr;
    char** vcf_hdr_off;
} gen_prsr_t;
    
typedef struct gen_cbcks_t
{
    ldoc_doc_t* (*proc_ln)(int fd, off_t mx, ldoc_doc_t* fdoc, ldoc_trie_t* idx, char* ln, size_t lnlen, gen_prsr_t* st, char** cmt, gen_fstat* stat);
} gen_cbcks_t;

typedef struct gen_ctxt_t
{
    gen_filetype_t tpe;
    bool py;
    bool qk;
    char* fgen;
    char* pycall;
    char* fname;
    FILE* fout;
    char* usr;
    char* ver;
} gen_ctxt_t;
    
void gen_err(int err, const char* s);
    
void gen_init();
    
void gen_nde_dsc_opt(ldoc_nde_t* nde, ldoc_nde_t* dsc, char* lbl);
    
char* gen_res_opt(ldoc_res_t* res);
char* gen_res_optx(ldoc_res_t* res);
char* gen_res_req(ldoc_res_t* res);
    
char* gen_term_crnl(char* s);
void gen_ky(char* attr, char** val);
void gen_lwr(char* str);
void gen_lwrhyph(char* str);
char* gen_quoskp(char* str);
void gen_kwd(char* str, gen_attr_t* attr, bi_attr upfail);
char gen_inv(char c);
char* gen_seq_enc(char* s);
char* gen_seq_dec(char* s, gen_filetype_t tpe);
    
void gen_csepstr_dup(ldoc_nde_t* dst, char* val, bool dup);
    
/**
 *  Returns the ordered-list node that was added to `dst`.
 */
ldoc_nde_t* gen_csep(ldoc_nde_t* dst, gen_attr_t kwd, char* ky, char* val);
ldoc_nde_t* gen_csep_dup(ldoc_nde_t* dst, gen_attr_t kwd, char* ky, char* val, bool dup);

void gen_xcig(char* str);
void gen_qk_revcig(char* str);
    
char* gen_escstr(char* str, gen_filetype_t tpe);
    
size_t gen_csplit(char* str, char c);

void gff_proc_tgt(ldoc_nde_t* nde, char* val);
    
ldoc_nde_t* gen_ctx(ldoc_nde_t* nde, bool* nw, const char* ctx);
ldoc_nde_t* gen_find_nde(ldoc_nde_t* ctnr1, ldoc_nde_t* ctnr2, char* ky);
void gen_add_nw(ldoc_nde_t* nde, ldoc_nde_t* usr);

ldoc_content_t gen_smrt_tpe(char* val);
ldoc_content_t gen_smrt_flttpe(char* val);
    
bool gen_join_nde(ldoc_nde_t* nde);
bool gen_join_attrs_key(char* id, ldoc_nde_t* nde, ldoc_ent_t* ent, char* attrs);
bool gen_join_attrs_ent(char* id, ldoc_ent_t* ent, char* attrs);
bool gen_join_attrs_nde(char* id, ldoc_nde_t* nde, char* attrs);
void gen_splt_attrs(ldoc_nde_t* ftr, ldoc_nde_t* usr, ldoc_nde_t* ref, ldoc_nde_t* vars, char* attrs, bi_attr upfail);

ldoc_nde_t* gen_variants(char* seq, char sep, char** vseqs, size_t* vnum);

void gen_rd(int fd, off_t mx, ldoc_trie_t* idx, gen_cbcks_t* cbcks, gen_ctxt_t* ctxt);

void gen_rd_doc(int fd, off_t mx, gen_ctxt_t* ctxt);
    
void gen_ser(gen_ctxt_t* ctxt, gen_ctpe_t ctpe, ldoc_doc_t* doc, ldoc_doc_t* opt, gen_fstat* stat);
    
//
// Genomics file formatting
//

void gen_proc_doc_prgm_kv(ldoc_nde_t* cntnr, char* ky, char* alt, char* sep);
bool gen_proc_doc_prgm(ldoc_nde_t* prgm, char* sep);
bool gen_proc_doc_usr(ldoc_nde_t* ftr);
bool gen_proc_nde(ldoc_nde_t* vars, char* attr, char* pre, char* astr, size_t vnum);
    
char* qk_alloc(size_t n);
void qk_free();
char* qk_heap_ptr();
char* qk_working_ptr();
bool qk_heap_empty();
void qk_purge();
char* qk_strdup(const char* s1);
char* qk_strndup(const char* s1, size_t n);
bool qk_strcat(const char* s1);
bool qk_strncat(const char* s1, size_t n);
    
#ifdef __cplusplus
} /* extern "C" */
#endif

#endif
